Source/AwsCommonRuntimeKit/io/Stream.swift

```diff
@@ -4,106 +4,80 @@
+    /// Seek the stream starting from beginning
+    case begin = 0
+    /// Seek the stream from End.
+    case end = 2
 }
 
-extension FileHandle: AwsStream {
+extension FileHandle: IStreamable {
+
     @inlinable
-    public var status: aws_stream_status {
-        return aws_stream_status(is_end_of_stream: self.length == self.offsetInFile, is_valid: true)
+    public func length() throws -> UInt64 {
+        let length: UInt64
+        let savedPos: UInt64
+        if #available(macOS 11, tvOS 13.4, iOS 13.4, watchOS 6.2, *) {
+            savedPos = try self.offset()
+            try self.seekToEnd()
+            length = try self.offset()
+        } else {
+            savedPos = self.offsetInFile
+            self.seekToEndOfFile()
+            length = self.offsetInFile
+        }
+        try self.seek(toOffset: savedPos)
+        return length
     }
 
     @inlinable
-    public var length: UInt {
-        let savedPos = self.offsetInFile
-        defer { self.seek(toFileOffset: savedPos ) }
-        self.seekToEndOfFile()
-        return UInt(self.offsetInFile)
+    public func seek(offset: UInt64) throws {
+        try self.seek(toOffset: offset)
     }
 
     @inlinable
-    public func seek(offset: Int64, basis: aws_stream_seek_basis) -> Bool {
-        let targetOffset: UInt64
-        if basis.rawValue == AWS_SSB_BEGIN.rawValue {
-            targetOffset = self.offsetInFile + UInt64(offset)
+    public func read(buffer: UnsafeMutableBufferPointer<UInt8>) throws -> Int {
+        let data: Data?
+        if #available(macOS 11, tvOS 13.4, iOS 13.4, watchOS 6.2, *) {
+            data = try self.read(upToCount: buffer.count)
         } else {
-            targetOffset = self.offsetInFile - UInt64(offset)
+            data = self.readData(ofLength: buffer.count)
         }
-        self.seek(toFileOffset: targetOffset)
-        return true
-    }
-
-    @inlinable
-    public func read(buffer: inout aws_byte_buf) -> Bool {
-        let data = self.readData(ofLength: buffer.capacity - buffer.len)
-        if data.count > 0 {
-            let result = buffer.buffer.advanced(by: buffer.len)
-            data.copyBytes(to: result, count: data.count)
-            buffer.len += data.count
-            return true
+        guard let data = data else {
+            return 0
         }
-        return !self.status.is_end_of_stream
-    }
-}
 
-private func doSeek(_ stream: UnsafeMutablePointer<aws_input_stream>!,
-                    _ offset: Int64,
-                    _ seekBasis: aws_stream_seek_basis) -> Int32 {
-    let inputStream = Unmanaged<AwsInputStream>.fromOpaque(stream.pointee.impl).takeUnretainedValue()
-    if inputStream.awsStream.seek(offset: offset, basis: seekBasis) {
-        return AWS_OP_SUCCESS
-    }
-    return AWS_OP_ERR
-}
+        guard let baseAddress = buffer.baseAddress else {
+            throw CommonRunTimeError.crtError(CRTError(code: Int32(AWS_IO_STREAM_READ_FAILED.rawValue)))
```

Comment:
just crash if `buffer.baseAddress` isn't set

would it be simpler if our `read()` API was more like [copyBytes(to:count:)](https://developer.apple.com/documentation/foundation/data/1780297-copybytes)? taking 2 separate pointer and count args? right now I see you warning users not to change `buffer.count` and then you have to check here that `buffer.baseAddress` is actually set.

Reply:
Thanks, fixed by https://developer.apple.com/documentation/foundation/data/1779913-copybytes.